hibernate-core/src/main/java/org/hibernate/query/hql/internal/SemanticQueryBuilder.java

```diff
@@ -2918,7 +2918,8 @@ public SqmPredicate visitExistsPredicate(HqlParser.ExistsPredicateContext ctx) {
 	@Override
 	public SqmPredicate visitBooleanExpressionPredicate(HqlParser.BooleanExpressionPredicateContext ctx) {
 		final SqmExpression<?> expression = (SqmExpression<?>) ctx.expression().accept( this );
-		if ( expression.getJavaType() != Boolean.class ) {
+		if ( expression.getExpressible() != null
+				&& expression.getExpressible().getExpressibleJavaType().getJavaType() != Boolean.class ) {
```

Comment:
```suggestion
		final SqmExpressible<?> expressible = expression.getExpressible();
		final JavaType<?> expressibleJavaType = expressible != null ? expressible.getExpressibleJavaType() : null;
		if ( expressibleJavaType == null || expressibleJavaType.getJavaType() != Boolean.class ) {
```

You might also create an utility method to resolve the expressible java type class for a `SqmExpression<?>`, since I believe we need to do the same thing in other places, something like: `SqmUtil#resolveExpressibleJavaTypeClass(SqmExpression<?> expression)`.